### ISSUE TITLE
Selecting SMSC GT from MprocRules

### ISSUE THREAD

**Issue by faizann** [RestComm/smscgateway#188]
Currently networkId defines the SMSC GT. This tightly couples the SS7 routing with SMSC GT itself as networkId is used for both. This means that GT A set for networkId 1 cannot be used with networkId 2 routing.

I suggest we allow setting of SCGT via Mproc rules. This way GT used for SRI/MTFSM can be decoupled from networkId. 
This has following advantages:
1-- Setting of GT in SMPP TLVs or HTTP parameters. 
2-- This also allows using bigger ranges of GT.
3-- SS7 route plan is less complicated .

This requires changing SriSbb in sendSri method and MtSbb for setupForwardShortMessageRequest. The method getServiceCenterAddressString
which gets SCGT via networkId can be made as a fallback, if no GT is set.

Same goes for HomeRouting SBB.

We will also need PreImsiMproc rule so that GT can be set before SRI is sent. 

Suggestions?

**Comment by vetss** [comment]
Hello @faizann 

can you please explain more this firstly:

> Same goes for HomeRouting SBB.
> We will also need PreImsiMproc rule so that GT can be set before SRI is sent.

Do you mean that you need somehow to force SRI requests to local HLR inside HR procedure with SMSC GW GT value that must be somehow configured by mproc rules ? If yes then depending of which staff this choice of GT value can be done ?

"Imsi trigger" is triggered after HLR resonse from a local HLR, not before.

**Comment by vetss** [comment]
Generally "configring of SMSC GW GT" demands code update for both sending and mproc rules parts. For example a solution like SmsSet carries a "desired GT" (set by mproc rules) and this overrides a configured GT for a networiId.

For HR, not clear for me, waiting for comments.

**Comment by faizann** [comment]
Hi Sergey

Indeed for MT SMS the GT can be carried as part of the SmsSet and which can override the configured GT. 

In HomeRouting the inbound SRI creates an outbound SRI. In this case just before outbound SRI is performed we could also change the GT for outbound SRI as well as networkId. This specific part gives more control on where SRI is sent out. Hence a PreImsiMproc or PreSriMproc is needed here.
We could also use this part and override things like generated IMSI's MNC/MCC, SkipSRI parameter. This would give control more to MProc rule and allows more routing control.

Since we are on the topic, I would also like to point out that sometimes the SMSC GT in MAP layer can be different than the GT in SCCP layer. This is sometimes useful when the outbound SRI/MTFSM in HomeRouting requires identifying the originating SMSC as that information is usually lost due to HomeRouting proxying all.

**Comment by vetss** [comment]
Hello @faizann 

for home routing we need in this case another new trigger ( PreImsi  ?) to be implemented. Also I do not fully understand what exactly "IMSI's MNC/MCC, SkipSRI" means. As for different GTs in SCCP / MAP levels it also makes sense in mproc rule processing.

@faizann do you mean you can contribute this case ? If yes we need to prepare a design for a solution.

**Comment by faizann** [comment]
Yeah this is only for HomeRouting as there is no Mproc between inbound-outbound SRI and hence little control.

The Generated/Fake IMSI in HR uses MNC/MCC from a memory loaded table of MNC/MCC based on prefixes. This is not always accurate and other systems with actual NumberPortability access can accurately give the MNC/MCC. This is interesting for billing purposes. So, if our MProc before outbound SRI can also override MNC/MCC then it helps.

SkipSRI is not such a major thing needed in mproc but as you know SRI Skipping boolean is attached to destination networkId and having control on when to skip and when not to skip from mproc itself could give more intelligence to the system.

I think networkId is a great idea but if we can make the general parameters of SMS more configurable by Mproc rules then we will get more control and routing will get less complicated. Leaving networkId more to JSS7. But this is a very long road.

I am available to help where ever I can.

**Comment by vetss** [comment]
ok @faizann I see your point.

Can you contribute the task so we can have it onboard. I can help you in a solution design if needed.

**Comment by faizann** [comment]
Hi

Yeah a solution from you would be very helpful. I will then contribute the code.

Thanks

**Comment by faizann** [comment]
Hi

I have worked out a simple solution using SmsSet but I am going to work on creating PreImsiRequest MProc.

**Comment by vetss** [comment]
Hello @faizann 

ok

**Comment by darkhacknet** [comment]
Can someone please help me to get my Local GT ? I cant find precise information about it. 
thanks guys !
